In [8]:
print("test")

test


In [ ]:
import pandas as pd
from ast import literal_eval # literal : 반복관련
import os

In [ ]:
# os.chdir(r'C:\01JSPARK\03python\02code\02content')
# Panda가 파일을 읽을때, 10번인덱스에 서로다른 데이터 타입이 섞여 있어서 에러발생
movies_metadata = pd.read_csv("data/movies_metadata.csv")
movies_metadata.shape

C:\Users\C414\AppData\Local\Temp\ipykernel_19012\2366159999.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies_metadata = pd.read_csv("data/movies_metadata.csv")


(45466, 24)

In [12]:
movies_metadata[10:11]


adult belongs_to_collection    budget  \
10  False                   NaN  62000000   

                                               genres homepage    id  \
10  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...      NaN  9087   

      imdb_id original_language          original_title  \
10  tt0112346                en  The American President   

                                             overview  ... release_date  \
10  Widowed U.S. president Andrew Shepherd, one of...  ...   1995-11-17   

        revenue runtime                          spoken_languages    status  \
10  107879496.0   106.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                              tagline                   title  \
10  Why can't the most powerful man in the world h...  The American President   

    video vote_average vote_count  
10  False          6.5      199.0  

[1 rows x 24 columns]

In [10]:
movies_metadata.head()
movies_metadata['genres'].head()

0    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
1    [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...
2    [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
3    [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...
4                       [{'id': 35, 'name': 'Comedy'}]
Name: genres, dtype: object

In [15]:
print(movies_metadata['genres'][0])

[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]


In [16]:
movies_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [17]:
links_small = pd.read_csv("data/links_small.csv")
links_small.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9125 entries, 0 to 9124
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  9125 non-null   int64  
 1   imdbId   9125 non-null   int64  
 2   tmdbId   9112 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 214.0 KB


In [18]:
links_small.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [21]:
# ID를 float => Int로 변경
# 결손치 tmbdId 결손치 13개
links_small["tmdbId"].notnull().count

<bound method Series.count of 0       True
1       True
2       True
3       True
4       True
        ... 
9120    True
9121    True
9122    True
9123    True
9124    True
Name: tmdbId, Length: 9125, dtype: bool>

In [22]:
links_small['tmdbId'].isnull().count

<bound method Series.count of 0       False
1       False
2       False
3       False
4       False
        ...  
9120    False
9121    False
9122    False
9123    False
9124    False
Name: tmdbId, Length: 9125, dtype: bool>

In [ ]:
links_small = links_small[links_small["tmdbId"].notnull()]["tmdbId"].astype('int')
links_small.shape

In [26]:
links_small.info()

<class 'pandas.core.series.Series'>
Index: 9112 entries, 0 to 9124
Series name: tmdbId
Non-Null Count  Dtype
--------------  -----
9112 non-null   int64
dtypes: int64(1)
memory usage: 142.4 KB


In [27]:
links_small.shape

(9112,)

In [29]:
movies_metadata.shape

(45466, 24)

In [30]:
movies_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [34]:
# links_small 숫자 : movies_metadata 는 object(string)
# movies_metadata['id'].isin(links_small.astype('str'))
movies_metadata_small = movies_metadata[movies_metadata['id'].isin(links_small.astype('str'))]
movies_metadata_small.shape

(9099, 24)

In [36]:
# 필요한 column 추출
# movies = movies_data_small['title'] => title만 movies로 이동
movies = movies_metadata_small[['title','popularity','genres', 'release_date']]
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30  
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15  
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22  
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22  
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10

In [ ]:
# 
print(movies[0:1])
print(movies.head(1))

       title popularity                                             genres  \
0  Toy Story  21.946943  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   

  release_date  
0   1995-10-30  
       title popularity                                             genres  \
0  Toy Story  21.946943  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   

  release_date  
0   1995-10-30  


In [ ]:
#data frame은 loc . iloc로 index search 문법이 다름 cf) numpy는 movies[0]
print(movies.iloc[0])
print(movies.iloc[0]['genres'])
print('-------------------------------------')
print(movies.loc[0,'title'])
print(movies.loc[0,['title','genres']])

title                                                   Toy Story
popularity                                              21.946943
genres          [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
release_date                                           1995-10-30
Name: 0, dtype: object
[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
-------------------------------------
Toy Story
title                                             Toy Story
genres    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
Name: 0, dtype: object


In [47]:
movies['genres'].loc[0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [48]:
# data frame
# apply 
df = pd.DataFrame({
    "math":[80,90,100],
    "eng":[75,85,95]
})

In [49]:
df

math  eng
0    80   75
1    90   85
2   100   95

In [50]:
df.apply(lambda x:x.mean())

math    90.0
eng     85.0
dtype: float64

### [{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]
### -> ['Animation','Comedy','Family']
### -> Animation Comedy Family

In [60]:
movies['genres'].eq('[]')
movies['genres'].eq('[]').value_counts()
movies['genres'].iloc[1]

"[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10751, 'name': 'Family'}]"

In [ ]:
# apply list => join

In [ ]:
# fillna : 문자로 되어있는 dictionary 를 한행씩 X 값으로 보내서 name value값을 list로 만듦
# movies['genres'].fillna('[]').apply[literal_eval].apply(lambda x: 참 if 조건 else 거짓 )
# movies['genres'].fillna('[]').apply[literal_eval].apply(lambda x: sorted( 표현식 for i in x ) if isinstance(x, list) else [] )
# test_list = movies['genres'].fillna('[]').apply[literal_eval].apply(lambda x: sorted( i['name'] for i in x ) if isinstance(x, list) else [] )
test_list = movies['genres'].fillna('[]')\
.apply[literal_eval]\
.apply(lambda x: sorted( i['name'] for i in x ) if isinstance(x, list) else [] )\
.apply(True if c else False)

TypeError: 'method' object is not subscriptable

In [67]:
# test_list = movies['genres'].fillna('[]').apply(literal_eval).apply(
#     lambda x: sorted([i['name'] for i in x]) if isinstance(x, list) else []
# )

movies['str_genres'] = movies['genres'].fillna('[]').apply(literal_eval).apply(
    lambda x: sorted([i['name'] for i in x]) if isinstance(x, list) else []
).apply(lambda x:" ".join(x) if len(x)>0 else None)

C:\Users\C414\AppData\Local\Temp\ipykernel_19012\1401783986.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['str_genres'] = movies['genres'].fillna('[]').apply(literal_eval).apply(


- genres에서 [] 데이터는 제외
 ### test_list = movies['genres'].fillna('[]')
- 문자열을 리스트데이터로 변경
 ### apply(literal_eval)
- 장르 데이터를 추출하여 리스트로 변경
 ### apply(lambda x: sorted([i['name'] for i in x]) if isinstance(x, list) else [])
- 리스트데이터를 하나의 문자열로 변환
 ### apply(lambda x:" ".join(x) if len(x)>0 else None)

In [69]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9099 entries, 0 to 45265
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   title         9099 non-null   object
 1   popularity    9099 non-null   object
 2   genres        9099 non-null   object
 3   release_date  9099 non-null   object
 4   str_genres    9064 non-null   object
dtypes: object(5)
memory usage: 684.6+ KB


In [70]:
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22   
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10   

                 str_genres  
0   Animation Comedy Family  
1  Adventure Family Fantasy  
2            Comedy Romance  
3      Comedy Drama Romance  
4                    Comedy

In [71]:
movies[['genres','str_genres']].head()

genres                str_genres
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   Animation Comedy Family
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...  Adventure Family Fantasy
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...            Comedy Romance
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...      Comedy Drama Romance
4                     [{'id': 35, 'name': 'Comedy'}]                    Comedy

In [72]:
# 일자 : 데이터타입으로 변경
movies['release_date'] = pd.to_datetime(movies['release_date'])

C:\Users\C414\AppData\Local\Temp\ipykernel_19012\3545074752.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['release_date'] = pd.to_datetime(movies['release_date'])


In [73]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9099 entries, 0 to 45265
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   title         9099 non-null   object        
 1   popularity    9099 non-null   object        
 2   genres        9099 non-null   object        
 3   release_date  9099 non-null   datetime64[ns]
 4   str_genres    9064 non-null   object        
dtypes: datetime64[ns](1), object(4)
memory usage: 684.6+ KB


In [74]:
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22   
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10   

                 str_genres  
0   Animation Comedy Family  
1  Adventure Family Fantasy  
2            Comedy Romance  
3      Comedy Drama Romance  
4                    Comedy

In [75]:
# release date에서 년도만 뽑아서 year 컬럼 생성
movies['year'] = movies['release_date'].dt.year
movies[['release_date', 'year']]

C:\Users\C414\AppData\Local\Temp\ipykernel_19012\2300885992.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['year'] = movies['release_date'].dt.year


release_date  year
0       1995-10-30  1995
1       1995-12-15  1995
2       1995-12-22  1995
3       1995-12-22  1995
4       1995-02-10  1995
...            ...   ...
40224   2016-07-29  2016
40503   2016-09-15  2016
44821   2000-07-08  2000
44826   2001-07-06  2001
45265   2014-08-15  2014

[9099 rows x 2 columns]

In [76]:
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22   
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10   

                 str_genres  year  
0   Animation Comedy Family  1995  
1  Adventure Family Fantasy  1995  
2            Comedy Romance  1995  
3      Comedy Drama Romance  1995  
4                    Comedy  1995

In [77]:
### 데이터 결측치 제거 처리
movies.isnull().sum()



title            0
popularity       0
genres           0
release_date     0
str_genres      35
year             0
dtype: int64

In [78]:
movies.shape

(9099, 6)

In [ ]:
# 결측치  : str_genres 비율 null수 / 전체
( movies.isnull().sum() / movies.shape[0] )

title           0.000000
popularity      0.000000
genres          0.000000
release_date    0.000000
str_genres      0.003847
year            0.000000
dtype: float64

In [ ]:
movies1 = movies.dropna()
movies1.head()


title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22   
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10   

                 str_genres  year  
0   Animation Comedy Family  1995  
1  Adventure Family Fantasy  1995  
2            Comedy Romance  1995  
3      Comedy Drama Romance  1995  
4                    Comedy  1995

In [83]:
# 결측치를 날리고 날아간 데이터의 index를 재조정해줘야함 reset index
movies = movies.dropna()
movies = movies.reset_index(drop=True)

In [84]:
movies.isnull().sum()

title           0
popularity      0
genres          0
release_date    0
str_genres      0
year            0
dtype: int64

### BoW기반 Contents Based Filtering

In [85]:
from sklearn.feature_extraction.text import CountVectorizer

In [86]:
bow_vector = CountVectorizer()

In [87]:
movies['str_genres']

0                                 Animation Comedy Family
1                                Adventure Family Fantasy
2                                          Comedy Romance
3                                    Comedy Drama Romance
4                                                  Comedy
                              ...                        
9059        Action Adventure Drama Horror Science Fiction
9060                                    Documentary Music
9061            Action Adventure Animation Family Fantasy
9062    Adventure Animation Family Fantasy Science Fic...
9063                                         Comedy Drama
Name: str_genres, Length: 9064, dtype: object

In [88]:
genre_mat = bow_vector.fit_transform(movies['str_genres'])

In [89]:
bow_vector.vocabulary_

{'animation': 2,
 'comedy': 3,
 'family': 7,
 'adventure': 1,
 'fantasy': 8,
 'romance': 16,
 'drama': 6,
 'action': 0,
 'crime': 4,
 'thriller': 18,
 'horror': 12,
 'history': 11,
 'science': 17,
 'fiction': 9,
 'mystery': 15,
 'war': 20,
 'foreign': 10,
 'music': 14,
 'documentary': 5,
 'western': 21,
 'tv': 19,
 'movie': 13}

In [90]:
genre_mat.shape

(9064, 22)

In [ ]:
sorted(bow_vector.vocabulary_.items())

[('action', 0),
 ('adventure', 1),
 ('animation', 2),
 ('comedy', 3),
 ('crime', 4),
 ('documentary', 5),
 ('drama', 6),
 ('family', 7),
 ('fantasy', 8),
 ('fiction', 9),
 ('foreign', 10),
 ('history', 11),
 ('horror', 12),
 ('movie', 13),
 ('music', 14),
 ('mystery', 15),
 ('romance', 16),
 ('science', 17),
 ('thriller', 18),
 ('tv', 19),
 ('war', 20),
 ('western', 21)]

In [92]:
dict(sorted(bow_vector.vocabulary_.items()))

{'action': 0,
 'adventure': 1,
 'animation': 2,
 'comedy': 3,
 'crime': 4,
 'documentary': 5,
 'drama': 6,
 'family': 7,
 'fantasy': 8,
 'fiction': 9,
 'foreign': 10,
 'history': 11,
 'horror': 12,
 'movie': 13,
 'music': 14,
 'mystery': 15,
 'romance': 16,
 'science': 17,
 'thriller': 18,
 'tv': 19,
 'war': 20,
 'western': 21}

In [93]:
genre_mat.info()

AttributeError: 'csr_matrix' object has no attribute 'info'

In [ ]:
# csr_matrix : info 불가 type은 가능
type(genre_mat)

scipy.sparse._csr.csr_matrix

In [96]:
genre_mat.toarray()

array([[0, 0, 1, ..., 0, 0, 0],
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [1, 1, 1, ..., 0, 0, 0],
       [0, 1, 1, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], shape=(9064, 22))

In [97]:
genre_mat.toarray()[0]

array([0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [98]:
movies.head()

title popularity  \
0                    Toy Story  21.946943   
1                      Jumanji  17.015539   
2             Grumpier Old Men    11.7129   
3            Waiting to Exhale   3.859495   
4  Father of the Bride Part II   8.387519   

                                              genres release_date  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   1995-12-22   
4                     [{'id': 35, 'name': 'Comedy'}]   1995-02-10   

                 str_genres  year  
0   Animation Comedy Family  1995  
1  Adventure Family Fantasy  1995  
2            Comedy Romance  1995  
3      Comedy Drama Romance  1995  
4                    Comedy  1995

In [99]:
genre_mat.toarray()[1]

array([0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

### 코사인 유사도 측정

In [ ]:
# genre_str로 volcabulary 22개의 vector를 만들어 유사도를 구함
# gpt 임베딩 차원수는 모델수마다 다르나 1000개이상의 차원을 지원함

In [100]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
similarity_of_genre = cosine_similarity(genre_mat, genre_mat) # data:genre_mat

In [102]:
similarity_of_genre.shape

(9064, 9064)

In [106]:
print(similarity_of_genre)

[[1.         0.33333333 0.40824829 ... 0.51639778 0.47140452 0.40824829]
 [0.33333333 1.         0.         ... 0.77459667 0.70710678 0.        ]
 [0.40824829 0.         1.         ... 0.         0.         0.5       ]
 ...
 [0.51639778 0.77459667 0.         ... 1.         0.73029674 0.        ]
 [0.47140452 0.70710678 0.         ... 0.73029674 1.         0.        ]
 [0.40824829 0.         0.5        ... 0.         0.         1.        ]]


In [ ]:
# 첫번째와 두번째 영화의 유사도는 0.33333333 

In [107]:
import numpy as np

In [ ]:
#list argsort : index로 sort 처리
a = np.array([30,40,51,47,40])
index = np.argsort(a)
index

array([0, 1, 4, 3, 2])

In [109]:
movie0_similarity = similarity_of_genre[0]
movie0_similarity

array([1.        , 0.33333333, 0.40824829, ..., 0.51639778, 0.47140452,
       0.40824829], shape=(9064,))

In [110]:
movies[['title', 'str_genres']].iloc[0:2,:]

title                str_genres
0  Toy Story   Animation Comedy Family
1    Jumanji  Adventure Family Fantasy

### 유사도가 높은 영화의 index로 정렬

In [114]:
sorted_similarity_of_genre = similarity_of_genre.argsort()
sorted_similarity_of_genre

array([[  38, 7848, 9060, ..., 2718, 7064,    0],
       [ 926, 9054, 9063, ..., 8327, 5010,    1],
       [  58,   73, 9059, ...,    6,   49,   60],
       ...,
       [9060,   60,   61, ..., 9061, 3358, 7904],
       [9040, 9054, 9060, ..., 3020, 4340, 4238],
       [3400, 9042, 9043, ..., 6537, 3404, 3394]], shape=(9064, 9064))

In [115]:
# index를 역순배열
sorted_similarity_of_genre = sorted_similarity_of_genre[:,::-1]
sorted_similarity_of_genre

array([[   0, 7064, 2718, ..., 9060, 7848,   38],
       [   1, 5010, 8327, ..., 9063, 9054,  926],
       [  60,   49,    6, ..., 9059,   73,   58],
       ...,
       [7904, 3358, 9061, ...,   61,   60, 9060],
       [4238, 4340, 3020, ..., 9060, 9054, 9040],
       [3394, 3404, 6537, ..., 9043, 9042, 3400]], shape=(9064, 9064))

In [116]:
# toy story와 장르 유사도 높은 영화
movies[['title', 'str_genres']].iloc[[0,7064],:]

title               str_genres
0                      Toy Story  Animation Comedy Family
7064  A Matter of Loaf and Death  Animation Comedy Family

In [119]:
# 쥬만지와 장르 유사도 높은 영화
movies[['title', 'str_genres']].iloc[[1,5010],:]

title                str_genres
1       Jumanji  Adventure Family Fantasy
5010  Peter Pan  Adventure Family Fantasy